### PR TITLE
trying to fix subsetting of the NWM nc files

### PR DESCRIPTION
**Pull request by jzwart**
We were getting weird values when reading the streamflow data from the subsetted nc files. We looked back to the subset_nwm.R script and think we fixed part of the problem. The streamflow dimensions were being screwed up when put into the new_nc files. We now loop through the sites (and ref time if forecast) and use ncvar_put to insert streamflow data into new_nc as well as multiply by the scale factor. If not multiplied by the scale factor when data is put into the nc, the data read out will be 100x's too small. Double check our looping - we think we mirrored what you had but couldn't test it. 

There were also 4 repeat COMIDS in the feature_id dimension. We're not sure why and did not know how to fix that. This is a secondary issue.

### CODE REVIEW

**Comment by dblodgett-usgs** [comment]
agh. I knew there would be a bug or two in what I did... I'll have a look tonight and get back to you.

File: 1_data/src/subset_nwm.R [hunk 1]
```diff
@@ -99,7 +99,27 @@ subset_nwm <- function(ind_file, model_configuration, comids, gd_config) {
...
   }
 
-  ncvar_put(new_nc, new_nc$var$streamflow, streamflow)
+  # need to loop through the sites because the dimensions in streamflow matrix were written incorrectly
+  for(s in 1:length(site_inds)) {
+    if(length(dimids) == 2) {
+      # Note axis order is assumed here!!!
+      ncvar_put(new_nc,
+                new_nc$var$streamflow,
+                streamflow[,s]*(1/new_nc$var$streamflow$scaleFact),  # need to multiply streamflow data by scale factor in nc var
```
Comment (dblodgett-usgs) [inline 1_data/src/subset_nwm.R]: Ugh... does ncdf4 apply the scale_factor on read but not on write?

**Comment by aappling-usgs** [comment]
thanks, @dblodgett-usgs ! i hope you're having a good trip

**Comment by jzwart** [comment]
yes! thanks @dblodgett-usgs !

File: 1_data/src/subset_nwm.R (hunk 1 above)
Comment (aappling-usgs) [reply 1_data/src/subset_nwm.R]: It seems that way...not at all intuitive, and it'd be great if you could test that with the nc data straight from THREDDS, but with our made-up data this seemed to be necessary in just one direction. The help file for ncvar_get says,
> If the variable in the netCDF file has a scale and/or offset attribute defined, the returned data are automatically and silently scaled and/or offset as requested.

while the help file for ncvar_put doesn't mention "scale" at all. https://cran.r-project.org/web/packages/ncdf4/ncdf4.pdf

Looks like another option might be to use raw_datavals=TRUE` in the call to ncvar_get...but I just noticed that, and I think the rescaling option we implemented above does work. I hope.

Comment (jzwart) [reply 1_data/src/subset_nwm.R]: It seems like it's applied on write but not read?? what we put in was scaled by the factor and before closing nc, reading with ncvar_get() showed a scaled version of what we ncvar_put()

File: 1_data/src/subset_nwm.R [hunk 2]
```diff
@@ -99,7 +99,27 @@ subset_nwm <- function(ind_file, model_configuration, comids, gd_config) {
     }
   }
 
-  ncvar_put(new_nc, new_nc$var$streamflow, streamflow)
+  # need to loop through the sites because the dimensions in streamflow matrix were written incorrectly
```
Comment (dblodgett-usgs) [inline 1_data/src/subset_nwm.R]: I'm not really following what is meant by "written incorrectly". We should be able to construct an array that can be written all in one go with ncvar_put right?

**Comment by dblodgett-usgs** [comment]
OK, so there are potentially two issues here. One that is really rough related to scale factors and ncdf4 and another that I'm not 100% understanding related to the way dimensions were written (I've not addressed this issue in this comment).

The issue with ncdf4's handling of scale_factor may go deeper than just this script. When processing all this data, I used this: 
`ncvar_put(new_nc, new_nc$var$streamflow, ncvar_get(nc, nc$var$streamflow)[keep])`

If `ncvar_get` applies scale factor and `ncvar_put` does not, then (FML) the actual streamflow values get written to an integer netcdf variable -- which will cull precision to the nearest whole CMS!  

That is with the first script I ran that is not in this repo. The next pass, in this repo, would be reading the data and applying the scale factor so you are getting data adjusted by a scale factor that should not be adjusted. -- I think. 

I am rebuilding with the code you checked in now for the sake of doing it. Will look at this a bit more later on and get a PR in for my updated build stuff.

**Comment by dblodgett-usgs** [comment]
Serious 😿 over here. 

I create this netcdf:
```
netcdf test_scale_factor {
dimensions:
	t = 1 ;
variables:
	int test(t) ;
		test:scale_factor = 0.1f ;
		test:units = "test_units" ;
data:

	test = 1 ;
}
```

I do:
```
> nc <- ncdf4::nc_open("test_scale_factor.nc", write = T)
> read_in <- ncdf4::ncvar_get(nc, "test")
> print(read_in)
[1] 0.1
```
So it applies the scale factor on the way in.  
Now I do:
```
> ncdf4::ncvar_put(nc, "test", read_in)
> nc_close(nc)
> quit()
$ ncdump test_scale_factor.nc 
netcdf test_scale_factor {
dimensions:
	t = 1 ;
variables:
	int test(t) ;
		test:scale_factor = 0.1f ;
		test:units = "test_units" ;
data:

 test = 0 ;
}
```
Like I said, FML.

If I use: `read_in <- ncdf4::ncvar_get(nc, "test", raw_datavals = TRUE)` -- I get the raw data and things are 1:1 on round trip. So... this is a problem. I need to get back to Tuscaloosa and rerun the full extract for this stuff. 

I still need to unravel the dimensions issue -- am getting errors with the code in this PR as well as my attempt to revert them. But I think that's secondary to the scaling issue.

**Comment by dblodgett-usgs** [comment]
The more I investigate this, the more I realize I need to do a thorough pass of my processing workflow from source to end product focusing on the scale factor attribute. I was assuming that R didn't mess with it which had impacts all the way through. At this stage, my hope is that you all can make progress without this data from me. I will get to this when back in the building in Tuscaloosa. 

I'm just going to make a bigger mess if I attempt to fix this from here.... sorry.

**Comment by dblodgett-usgs** [comment]
OK, last comment for tonight. I have no idea what to say about the operational output. I really need to look at the source data that I don't have in Tuscaloosa to know exactly what happened with the scale factor. The streamflow variable in my local data is missing the scale_factor attribute... which has me worried. I never touched it in my work and I always write attributes from input to output. I'm not going to speculate on what happened to it. I need to step through my processing and see. I think I can do that remotely, but it's not going to help since I can't reprocess in time for you to use the output.

The retrospective data you have in those netCDF files should be OK if you DON'T apply the scale factor? However, you will only get whole CMS -- the decimals have been chopped off by this scale_factor bug.

**Pull request merged**